### DISCUSSION TITLE
Gathering DOFs on CUDA

### DISCUSSION THREAD

**Discussion by Marshdweller** [Nek5000/nekRS#560] [Q&A] [answered]
Hi everyone,

I have been studying the GPU kernels of NekRS for elliptic equations, specifically focusing on the axhelm kernel. From my understanding, the matrix-vector multiplication is performed on the separated degrees of freedom (DOFs) per element. There should be an assembly ("gathering") phase where duplicated DOFs are accumulated into physical DOFs. I believe you call it "gather-scatter operator". In classical Nek, it seems to be in `oogs`, as suggested 
[Here](https://github.com/Nek5000/Nek5000/issues/832) by @yslan 

However, in NekRS I have been unable to locate a CUDA kernel where this gathering and scattering occur. Could someone please point me in the right direction?

Thank you!

**Comment by yslan** [comment, marked as answer]
For nekrs v23, the `ogs` is put under gslib (gather scatter library)
https://github.com/Nek5000/nekRS/tree/master/3rd_party/gslib/ogs

We use OCCA as an abstract layer for all device/CPU backends so you won't find CUDA kernel in the source code.

See also 3.2 parallel communication in this [paper](https://www.sciencedirect.com/science/article/pii/S0167819122000710) for different communication protocol.
